### PR TITLE
[Push] Sync subscriptions

### PR DESCRIPTION
**Pull request by flypaper0**
# Description

Resolves # (issue)

## Due Dilligence

* [ ] Breaking change
* [ ] Requires a documentation update

### CODE REVIEW

File: Sources/WalletConnectPush/Client/Common/DeletePushSubscriptionSubscriber.swift [hunk 1]
```diff
@@ -31,7 +29,9 @@ class DeletePushSubscriptionSubscriber {
                 logger.debug("Peer deleted subscription")
                 let topic = payload.topic
                 networkingInteractor.unsubscribe(topic: topic)
-                pushSubscriptionStore.delete(forKey: topic)
+                Task(priority: .high) {
+                    try await pushSubscriptionStore.delete(id: topic)
+                }
```
Comment (llbartekll) [inline Sources/WalletConnectPush/Client/Common/DeletePushSubscriptionSubscriber.swift:34]: so if client 1A receives delete request from client 2
1A will publish sync delete 
then 1B will receive delete request from client 2 too and then sync delete from 1A?
so 1B will remove the subscription and publish sync delete too?
and then 1A receives sync delete from 1B for already removed subscription?

File: Sources/WalletConnectPush/Client/Wallet/ProtocolEngine/wc_notifyUpdate/NotifyUpdateResponseSubscriber.swift [hunk 2]
```diff
@@ -50,7 +50,7 @@ class NotifyUpdateResponseSubscriber {
                     
                     let updatedSubscription = PushSubscription(topic: subscriptionTopic, account: oldSubscription.account, relay: oldSubscription.relay, metadata: oldSubscription.metadata, scope: scope, expiry: expiry)
 
-                    subscriptionsStore.set(updatedSubscription, forKey: subscriptionTopic)
+                    try await subscriptionsStore.set(object: updatedSubscription, for: updatedSubscription.account)
```
Comment (llbartekll) [inline Sources/WalletConnectPush/Client/Wallet/ProtocolEngine/wc_notifyUpdate/NotifyUpdateResponseSubscriber.swift:53]: why do we get for topic and set for account?

File: Sources/WalletConnectPush/Client/Wallet/WalletPushClient.swift [hunk 3]
```diff
@@ -84,10 +83,12 @@ public class WalletPushClient {
         self.notifyUpdateRequester = notifyUpdateRequester
         self.notifyUpdateResponseSubscriber = notifyUpdateResponseSubscriber
         self.notifyProposeResponder = notifyProposeResponder
+        self.subscriptionsStore = subscriptionsStore
         setupSubscriptions()
     }
 
     public func subscribe(metadata: AppMetadata, account: Account, onSign: @escaping SigningCallback) async throws {
+        try await subscriptionsStore.initialize(for: account)
```
Comment (llbartekll) [inline Sources/WalletConnectPush/Client/Wallet/WalletPushClient.swift:91]: so from here I understand that one store can initialise multiple sync services, correct?

File: Sources/WalletConnectPush/PushStorageIdntifiers.swift [hunk 4]
```diff
@@ -1,7 +1,8 @@
 import Foundation
 
 enum PushStorageIdntifiers {
-    static let pushSubscription = "com.walletconnect.sdk.pushSbscription"
+    static let pushSubscription = "com.walletconnect.push.subscription"
```
Comment (llbartekll) [inline Sources/WalletConnectPush/PushStorageIdntifiers.swift:4]: why this has to be changed?

**Review by llbartekll — COMMENTED** [review]
Thanks for taking a look at it so fast 🙏
I understand that this adds sync store for threads but we also need to sync thread keys
and do the same for messages, correct?

File: Sources/WalletConnectPush/Client/Common/DeletePushSubscriptionSubscriber.swift (hunk 1 above)
Comment (flypaper0) [reply Sources/WalletConnectPush/Client/Common/DeletePushSubscriptionSubscriber.swift]: Sync delete only sends if state was changed (SyncStore line 78)

File: Sources/WalletConnectPush/Client/Wallet/ProtocolEngine/wc_notifyUpdate/NotifyUpdateResponseSubscriber.swift (hunk 2 above)
Comment (flypaper0) [reply Sources/WalletConnectPush/Client/Wallet/ProtocolEngine/wc_notifyUpdate/NotifyUpdateResponseSubscriber.swift]: account is using to get syncTopic (SyncStore line 68)

File: Sources/WalletConnectPush/Client/Wallet/WalletPushClient.swift (hunk 3 above)
Comment (flypaper0) [reply Sources/WalletConnectPush/Client/Wallet/WalletPushClient.swift]: yes, one store can work with multiple accounts. On `initialize` we're doing sync topic derivation and subscribing for sync topic

File: Sources/WalletConnectPush/PushStorageIdntifiers.swift (hunk 4 above)
Comment (flypaper0) [reply Sources/WalletConnectPush/PushStorageIdntifiers.swift:4]: to be consistent with another sync stores namings

**Comment by flypaper0** [comment]
> I understand that this adds sync store for threads but we also need to sync thread keys

yes, we're discussing keys sync here: https://github.com/WalletConnect/walletconnect-docs/pull/773/files

File: Example/WalletApp/ApplicationLayer/EthKeyStore.swift [hunk 5]
```diff
@@ -7,19 +7,22 @@ class EthKeyStore {
...
     var address: String {
-        return privateKey.address.hex(eip55: false)
+        return privateKey.address.hex(eip55: true)
     }
 
     var privateKeyRaw: Data {
         return Data(privateKey.rawPrivateKey)
     }
 
     private init() {
```
[the remark this answers is not in the document]
Comment (llbartekll) [reply Example/WalletApp/ApplicationLayer/EthKeyStore.swift:17]: idk🤔 extra screen on the first app launch? the one from showcase app?

File: Sources/WalletConnectPush/Client/Wallet/ProtocolEngine/wc_pushSubscribe/PushSubscribeResponseSubscriber.swift [hunk 6]
```diff
@@ -47,7 +43,8 @@ class PushSubscribeResponseSubscriber {
 
                     guard let responseKeys = kms.getAgreementSecret(for: payload.topic) else {
                         logger.debug("PushSubscribeResponseSubscriber: no symmetric key for topic \(payload.topic)")
-                        subscriptionPublisherSubject.send(.failure(Errors.couldNotCreateSubscription))
```
Comment (llbartekll) [inline Sources/WalletConnectPush/Client/Wallet/ProtocolEngine/wc_pushSubscribe/PushSubscribeResponseSubscriber.swift:50]: why commented out?

File: Sources/WalletConnectPush/Client/Wallet/ProtocolEngine/wc_pushSubscribe/PushSubscribeResponseSubscriber.swift [hunk 7]
```diff
@@ -76,26 +73,26 @@ class PushSubscribeResponseSubscriber {
                         try await networkingInteractor.subscribe(topic: pushSubscriptionTopic)
                     } catch {
                         logger.debug("PushSubscribeResponseSubscriber: error: \(error)")
-                        subscriptionPublisherSubject.send(.failure(Errors.couldNotCreateSubscription))
+                        // TODO: Error Subscription
+                        // subscriptionPublisherSubject.send(.failure(Errors.couldNotCreateSubscription))
```
Comment (llbartekll) [inline Sources/WalletConnectPush/Client/Wallet/ProtocolEngine/wc_pushSubscribe/PushSubscribeResponseSubscriber.swift:77]: same here

File: Sources/WalletConnectPush/Client/Wallet/ProtocolEngine/wc_pushSubscribe/PushSubscribeResponseSubscriber.swift [hunk 8]
```diff
@@ -76,26 +73,26 @@ class PushSubscribeResponseSubscriber {
...
                         return
                     }
 
                     guard let metadata = metadata else {
                         logger.debug("PushSubscribeResponseSubscriber: no metadata for topic: \(pushSubscriptionTopic!)")
-                        subscriptionPublisherSubject.send(.failure(Errors.couldNotCreateSubscription))
+                        // TODO: Error Subscription
```
Comment (llbartekll) [inline Sources/WalletConnectPush/Client/Wallet/ProtocolEngine/wc_pushSubscribe/PushSubscribeResponseSubscriber.swift:83]: and here

File: Sources/WalletConnectPush/Client/Wallet/WalletPushClient.swift [hunk 9]
```diff
@@ -56,35 +52,41 @@ public class WalletPushClient {
...
         self.pushSubscribeResponseSubscriber = pushSubscribeResponseSubscriber
+        self.deletePushSubscriptionSubscriber = deletePushSubscriptionSubscriber
         self.notifyUpdateRequester = notifyUpdateRequester
         self.notifyUpdateResponseSubscriber = notifyUpdateResponseSubscriber
         self.notifyProposeResponder = notifyProposeResponder
         self.notifyProposeSubscriber = notifyProposeSubscriber
-        self.deletePushSubscriptionSubscriber = deletePushSubscriptionSubscriber
+    }
+
+    public func register(account: Account, onSign: @escaping SigningCallback) async throws {
```
Comment (llbartekll) [inline Sources/WalletConnectPush/Client/Wallet/WalletPushClient.swift:86]: the method should be called enableSync according to specs
https://github.com/WalletConnect/walletconnect-docs/pull/773/files

File: Sources/WalletConnectPush/Client/Wallet/WalletPushClient.swift [hunk 10]
```diff
@@ -124,6 +126,23 @@ public class WalletPushClient {
...
+        switch result {
+        case .signed(let signature):
+            try await syncClient.register(account: account, signature: signature)
+            logger.debug("Sync pushSubscriptions store registered and initialized")
+        case .rejected:
+            throw PushError.registerSignatureRejected
+        }
+    }
+}
+
```
Comment (llbartekll) [inline Sources/WalletConnectPush/Client/Wallet/WalletPushClient.swift:145]: could be move it to a separate class to not pollute wallet client?
syncRegisterService or smth like that

**Review by llbartekll — APPROVED** [review]

File: Sources/WalletConnectPush/Client/Wallet/ProtocolEngine/wc_pushSubscribe/PushSubscribeResponseSubscriber.swift (hunk 6 above)
Comment (flypaper0) [reply Sources/WalletConnectPush/Client/Wallet/ProtocolEngine/wc_pushSubscribe/PushSubscribeResponseSubscriber.swift:50]: Added subscriptionErrorPublisher

File: Sources/WalletConnectPush/Client/Wallet/WalletPushClient.swift (hunk 10 above)
Comment (flypaper0) [reply Sources/WalletConnectPush/Client/Wallet/WalletPushClient.swift]: created PushSyncService

File: Example/IntegrationTests/Push/PushTests.swift [hunk 11]
```diff
@@ -21,10 +23,19 @@ final class PushTests: XCTestCase {
 
     var pairingStorage: PairingStorage!
 
+    let pk = try! EthereumPrivateKey()
```
Comment (llbartekll) [inline Example/IntegrationTests/Push/PushTests.swift:26]: sometimes we re running those tests against prod cast and prod gm dapp. Can we use some fixed entropy here for deriving the private key? to not pollute their DBs

**Review by llbartekll — APPROVED** [review]

File: Example/IntegrationTests/Push/PushTests.swift (hunk 11 above)
Comment (flypaper0) [reply Example/IntegrationTests/Push/PushTests.swift:26]: Why we can pollute db's? I think there will be same amount of relay messages in both cases (random and derived private key)

**Pull request merged**